### ISSUE TITLE
Keys should not be stored in azd environments or outputted in bicep

### ISSUE THREAD

**Issue by jongio** [Azure-Samples/azure-search-openai-demo#919]
This repo outputs the openAI key here: 
https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/infra/main.bicep#L392

That will add the secret to the Azure deployment and it will add the secret to the azd environment in clear text.

**Comment by jongio** [comment]
Consider using a secure input for the key #920 and then save that to KV and read that value from KV instead of putting in azd env.

**Comment by pamelafox** [comment]
Ah good call out. We have an upcoming PR (from a private fork) that adds KeyVault for a different key so it will be easier to use KV to store the OpenAI key after that is in.

**Comment by pamelafox** [comment]
@jongio To confirm, were you imagining the flow would be:

* azd env set OPENAI_KEY keyhere
* azd up
* main.bicep stores openAiKey in keyvault
* app code checks keyvault for key

OR:

* azd up
* main.bicep creates a keyvault
* user manually adds key to keyvault
* app code checks keyvault for key

The first flow has drawback that the secret will be stored in plaintext in .azure/ENVNAME/.env, but is easier for user.

And I specifically didn't suggest possibility of using App Service environment variables with KeyVault references since we want access to this secret locally as well.

**Comment by jongio** [comment]
> @jongio To confirm, were you imagining the flow would be:
> 
> * azd env set OPENAI_KEY keyhere
> * azd up
> * main.bicep stores openAiKey in keyvault
> * app code checks keyvault for key
> 
> OR:
> 
> * azd up
> * main.bicep creates a keyvault
> * user manually adds key to keyvault
> * app code checks keyvault for key
> 
> The first flow has drawback that the secret will be stored in plaintext in .azure/ENVNAME/.env, but is easier for user.
> 
> And I specifically didn't suggest possibility of using App Service environment variables with KeyVault references since we want access to this secret locally as well.

Right now, we can't recommend putting a secret in the azd env because the .env file could be inadvertently committed to GH.com.

I think, the best we can do is ask them to set the env var for the key in the current shell and then use it in bicep.  You can have secure inputs, just not secure outputs (yet).

**Comment by pamelafox** [comment]
If you do `azd env set`, doesn't that immediately update the env file? (I just tried, and it did). I'm trying to figure out how they'd set something for use in the Bicep file without using `azd env set`.

**Comment by jongio** [comment]
Yes, but we don't want to encourage people to store keys in azd env b/c clear text and only thing preventing it from getting committed is a .gitignore.

I believe the only thing we can do is tell them to set an environment variable manually with export or set

**Comment by mattmsft** [comment]
There's a bit of a catch-22 here right now

#1052 

There are some secrets that only exist in the .env file, but we need to move them to keyvault as part of resolving this issue
We can use "secretOrRandomPassword" (https://github.com/Azure/azure-dev/discussions/940) to generate a new secret, but we can't specify an optional "default" value in case the user has previously run azd up and has the secrets initialized already.

I'll have to think about this one for a bit - please let me know if you have any thoughts on how to move secrets from the .env file to key vault if they already exist

**Comment by jongio** [comment]
I wonder if we can use the same method that we use here with the default of USE_APIM

<img width="428" alt="image" src="https://github.com/Azure-Samples/azure-search-openai-demo/assets/2163001/fde574db-816b-430a-ba5d-c1e2bbdaf45b">


https://github.com/Azure-Samples/todo-csharp-sql/blob/main/infra/main.parameters.json
